Add smartcase to the regex engine.

jamiedsmith95 opened the pull request:
Implementation of smart case for regex, using (?c) as a modifier flag.
I believe completes feature request: https://github.com/mawww/kakoune/issues/4856

mawww commented on src/regex_impl.cc:
```diff
@@ -212,6 +211,7 @@ struct RegexParser
                     case 'I': m_flags &= ~Flags::IgnoreCase; break;
                     case 's': m_flags |= Flags::DotMatchesNewLine; break;
                     case 'S': m_flags &= ~Flags::DotMatchesNewLine; break;
+                    case 'c': m_flags |= Flags::SmartCase; m_flags |= Flags::IgnoreCase; break;
```
`m_flags |= Flags::SmartCase | Flags::IgnoreCase;` ?

mawww commented on src/regex_impl.cc:
```diff
@@ -328,6 +328,14 @@ struct RegexParser
                     parse_error(format("unexpected '{}'", cp));
                 ++m_pos;
+                if (is_upper(cp) && (m_flags & Flags::SmartCase)) {
+                  m_flags &= ~Flags::IgnoreCase;
```
Why do we clear IgnoreCase here ? Doesn't that mean we have a different behaviour for the first upper case character ?

jamiedsmith95 replied:
Shouldn't do, the IgnoreCase isn't checked until the add_node call just after this check.

mawww reviewed (changes requested):
A few comments on the implementation, would be nice to add a few test cases and add this to the regex documentation.

mawww commented on src/regex_impl.cc:
```diff
@@ -9,7 +8,6 @@
 #include "string_utils.hh"
 #include "vector.hh"
 #include "utils.hh"
-
```
Can we leave those empty lines ?

jamiedsmith95 commented:
> A few comments on the implementation, would be nice to add a few test cases and add this to the regex documentation.

I have addressed your comments. I have also modified it to also make an uppercase char alter existing character classes (otherwise the behaviour of character classes would depend on whether they were before or after the uppercase). A couple of tests and a line in the modifier part of the docs. The only thing to emphasise is that character classes will depend on smart case but will not themselves trigger it. e.g. [A-Z] will not count as an uppercase.

Supreeeme commented:
@jamiedsmith95 Are you still interested in getting this merged, or would you mind if I took it over?

d3m3t3r commented on src/regex_impl.cc:
```diff
@@ -334,8 +334,9 @@ struct RegexParser
                         node.ignore_case = false;
-                  
-
+                    for (CharacterClass &char_class : m_parsed_regex.character_classes)
```
This is insufficient as these character classes might have already been built as case-insensitive (see character_class() method using to to_lower() on ranges).  E.g. "(?c)[A-Z]+foO"  wouldn't match "BARfoO" because '[A-Z]' was already created as '[a-z]'.

d3m3t3r commented on src/regex_impl.cc:
```diff
@@ -334,8 +334,9 @@ struct RegexParser
                     m_flags &= ~Flags::IgnoreCase;
```
I believe SmartCase flag should be cleared too as it's not necessary to execute this block for every subsequent uppercase codepoint again and again.
